### ISSUE TITLE
SQL generation bug

### ISSUE THREAD

**Issue by szguoxz** [linq2db/linq2db#3636]
this bug happens on all provider (I tested on SQL server and PostgreSQL).
This is a weird bug. My coworker wrote a very long and complicated LinQ, and can never got the right result. (actually we got complaint from our client, so this is a painful experience) So I simplified the Linq, and eventually figureout there might be a bug with grouping.
The final Linq May not make sense to you, that's because I simplified it, and the you won't see the business logic in it any more. 

Anyway, here is the code to recreate the bug:
basically, the parameter passed into the final SQL is wrong the second time. 
When I call Test(2), it looks fine, but the finally SQL looks odd, even the result is correct. The finally SQL created two parameter for myid:  :myid, :ID2

This is fine, but when I call Test(85), funny things start to happen:
:myid = 2
:ID2 = 85
were passed in. But they both should be 85.

It's really odd, but I recreated the test case, and always got the same wrong result.
You can simply copy the following code, and run it, and see it yourself.

I print the parameter value using the intercepter.

```cs
using static System.Console;
using LinqToDB;
using LinqToDB.Mapping;
using LinqToDB.Data;
using LinqToDB.Interceptors;
using System.Data.Common;
AppDataContext db = new("PostgreSQL", "conn");
db.CreateTable<T1>();
db.CreateTable<T2>();

db.Insert(new T1 { ID = 1, ID2 = 2 });
db.Insert(new T1 { ID = 1, ID2 = 85 });
db.Insert(new T1 { ID = 2, ID2 = 2 });
db.Insert(new T1 { ID = 2, ID2 = 85 });

db.Insert(new T2 { ID = 1, ID2 = 2 });
db.Insert(new T2 { ID = 1, ID2 = 2 });

Test(2);
Test(85);


void Test(int myid)
{
	var q = from s in db.GetTable<T1>().Where(x => x.ID2 == myid)
			join o in db.GetTable<T2>().Where(x => x.ID2 == myid) on s.ID equals o.ID into temp1
			from order in temp1.DefaultIfEmpty()
			group s by s.ID into g
			select g;

	WriteLine(System.Text.Json.JsonSerializer.Serialize(q.FirstOrDefault()));
}

public partial class AppDataContext : DataConnection
{
	public AppDataContext(string providerName, string connectionString) : base(providerName, connectionString)	
	{
		AddInterceptor(new MyCMD());
	}

	class MyCMD : LinqToDB.Interceptors.CommandInterceptor
	{
		public override DbCommand CommandInitialized(CommandEventData eventData, DbCommand command)
		{
			Console.WriteLine(command.CommandText);
			foreach (DbParameter p in command.Parameters) Console.WriteLine(p.ParameterName + " : " + p.Value);
			return base.CommandInitialized(eventData, command);
		}
	}
}

[Table]
public class T1
{
	/// <summary>
	/// ID//
	/// </summary>
	[Column("id")] public int ID { get; set; } // integer
	[Column("id2")] public int ID2 { get; set; } // integer
	[Column("id3")] public int ID3 { get; set; } // integer
}


[Table]
public class T2
{
	/// <summary>
	/// ID//
	/// </summary>
	[Column("id")] public int ID { get; set; } // integer
	[Column("id2")] public int ID2 { get; set; } // integer
}
```

**Comment by MaceWindu** [comment]
As workaround you can try to clear query cache `Query.ClearCaches();` before or after this query. It will affect performance, but if you don't use a lot of queries - it is fine

**Comment by szguoxz** [comment]
I changed my query to work around it. :-)
There's two issues here:
1. apparently query cache issue.
2. I supplied one variable, myid, but the generated SQL generated two parameters: :myid, :ID2. I think this is unnecessary and might introduce unexpected result down the road. We really should stick to one parameter approach. (If the query is not too complicated, it uses one parameter in both places. But when it hits some painpoint, it becomes two parameters.) 

And the parameter names a re disturbing, "myid", "ID2".  It really should be @P1, @P2, so we can be sure there won't be any issues. Arbitrary names could cause issues.

**Comment by sdanyliv** [comment]
@szguoxz, one question. Do you want to select first element of the group?

**Comment by szguoxz** [comment]
Actually I want all. I did the test to reduce the data returned to me. This is the simplified Linq. Not my original Linq.
I just try to simplify the Linq as much as possible but reserve the bug. This is the simplest I can get. It tells you how to recreate the bug. The Linq itself does not have business meaning now.

I was hoping you can get to the bottom of this bug. Not just fix my SQL. I've already fixed my Linq using a different approach.
Seems like you guys fix it? I will get the latest code tonight and test it.

Is it stable enough to use in producton?

**Comment by sdanyliv** [comment]
Note that GroupBy followed by selecting grouped items has no sense and no direct SQL translation. There are other techniques to do that effectively.

**Comment by szguoxz** [comment]
> Note that GroupBy followed by selecting grouped items has no sense and no direct SQL translation. There are other techniques to do that effectively.

The test Linq is just a simplified way to show you how to reproduce the cache bug. The real Linq is too complicated for me to show it here.

**Comment by sdanyliv** [comment]
Problem was in materialization of GroupBy result. That’s why I warn that this is ineffective query.

**Comment by MaceWindu** [comment]
@szguoxz , 

1. we plan to revert fix for this issue here #3720
2. we cannot reproduce it anymore with current master

What I want you to ask is to retest your issue using preview version of linq2db from Azure artifacts feed  (after #3720 merged)

**Comment by MaceWindu** [comment]
Nevermind, we will introduce better fix in next release
